src/Prefetch.cpp

```diff
@@ -0,0 +1,280 @@
+            Function varf = varit->second;
+            if (varf.outputs()) {
+                vector<Type> varts = varf.output_types();
+                t = varts[0];
+                has_static_type = true;
+            }
+        }
+
+        return t;
+    }
+
+    // Generate the required prefetch code (lets and statements) for the
+    // specified varname and box
+    //
+    Stmt prefetch_box(const string &varname, const Box &box)
+    {
+        vector<pair<string, Expr>> plets;  // Prefetch let assignments
+        vector<Stmt> pstmts;               // Prefetch stmt sequence
+
+        int dims = box.size();
+        bool has_static_type = true;
+        Type t = get_type(varname, has_static_type);
+        string elem_size_name = varname + ".elem_size";
+        Expr elem_size_bytes;
+
+        if (has_static_type) {
+            elem_size_bytes = t.bytes();
+        } else {   // Use element size for inputs that don't have static types
+            Expr elem_size_var = Variable::make(Int(32), elem_size_name);
+            elem_size_bytes = elem_size_var;
+        }
+
+        std::ostringstream ss; ss << t;
+        string type_name = ss.str();
+        string pstr = unique_name('p');
+        debug(1) << "  prefetch #" << pstr << ": "
+                             << varname << " ("
+                             << (has_static_type ? type_name : elem_size_name)
+                             << ", dims:" << dims << ")\n";
+
+        // TODO: Opt: check box if it should be prefetched?
+        // TODO       - Only prefetch if varying by loop_var_name?
+        // TODO       - Don't prefetch if "small" all constant dimensions?
+        // TODO         e.g. see: camera_pipe.cpp corrected matrix(4,3)
+
+        string varname_prefetch_buf = varname + "_prefetch_" + pstr + "_buf";
+        Expr var_prefetch_buf = Variable::make(Int(32), varname_prefetch_buf);
+
+        // Establish the variables for buffer strides, box min & max
+        vector<Expr> stride_var(dims);
+        vector<Expr> extent_var(dims);
+        vector<Expr> min_var(dims);
+        vector<Expr> max_var(dims);
+        for (int i = 0; i < dims; i++) {
+            string istr = std::to_string(i);
+            // string extent_name = varname + ".extent." + istr;
+            string stride_name = varname + ".stride." + istr;
+            string extent_name = varname + ".extent." + istr;
+            string min_name = varname + "_prefetch_" + pstr + "_min_" + istr;
+            string max_name = varname + "_prefetch_" + pstr + "_max_" + istr;
+
+            stride_var[i] = Variable::make(Int(32), stride_name);
+            extent_var[i] = Variable::make(Int(32), extent_name);
+            min_var[i] = Variable::make(Int(32), min_name);
+            max_var[i] = Variable::make(Int(32), max_name);
+
+            // Record let assignments
+            // except for stride, already defined elsewhere
+            plets.push_back(make_pair(min_name, box[i].min));
```

Comment:
FYI, removing the min/max plets this seems to slow down HelloHexagon down ~5%...probably due to the fact that allowing the (max-min+1) expressions to expand & simplify is obfuscating or eliminating a CSE opportunity from LLVM.